### ISSUE TITLE
Cannot subscribe_room for user after ejabberd restart

### ISSUE THREAD

**Issue by arjun9** [processone/ejabberd#3799]
## Environment

- ejabberd version: <could not retrieve this>
- Erlang version: 11.0.3
- OS: Linux Ubuntu (Debian)
- Installed from: source

## Configuration (only if needed): grep -Ev '^$|^\s*#' ejabberd.yml

<details>

```yaml
##
###              ejabberd configuration file
###
### The parameters used in this configuration file are explained at
###
###       https://docs.ejabberd.im/admin/configuration
###
### The configuration file is written in YAML.
### *******************************************************
### *******           !!! WARNING !!!               *******
### *******     YAML IS INDENTATION SENSITIVE       *******
### ******* MAKE SURE YOU INDENT SECTIONS CORRECTLY *******
### *******************************************************
### Refer to http://en.wikipedia.org/wiki/YAML for the brief description.
###
hosts:
  - dev-pigeon.mycompany.com
loglevel: debug
## If you already have certificates, list them here
certfiles:
     - /etc/ssl/ejabberd/mycompany.cert
     - /etc/ssl/ejabberd/mycompany.pem
#certfiles:
#   	- /etc/ssl/ejabberd/mycompany.pem
#  - /etc/letsencrypt/live/domain.tld/privkey.pem
oom_watermark: 50
listen:
  -
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    max_stanza_size: 262144
    shaper: c2s_shaper
    access: c2s
    starttls_required: true
  -
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
    max_stanza_size: 524288
  -
    port: 8443
    ip: "::"
    module: ejabberd_http
    tls: true
    #certfile: "/etc/ssl/ejabberd/mycompany.cert"
    request_handlers:
      /websockets: ejabberd_http_ws
      /bosh: mod_bosh
      /admin: ejabberd_web_admin
  -
    port: 5281
    ip: "::"
    module: ejabberd_http
    request_handlers:
      /api: mod_http_api
  -
    port: 5280
    ip: "::"
    module: ejabberd_http
    request_handlers:
      /bosh: mod_bosh
      /websockets: ejabberd_http_ws
      /admin: ejabberd_web_admin
      # /.well-known/acme-challenge: ejabberd_acme
  -
    port: 3478
    transport: udp
    module: ejabberd_stun
    use_turn: true
    ## The server's public IPv4 address:
    # turn_ip: 203.0.113.3
  -
    port: 1883
    ip: "::"
    module: mod_mqtt
    backlog: 1000
s2s_use_starttls: optional
acl:
  admin:
    user:
      - "admin": "dev-pigeon.mycompany.com"
      - "bot_admin": "dev-pigeon.mycompany.com"
    ip: "0.0.0.0"
  local:
    user_regexp: ""
  loopback:
    ip:
      - "127.0.0.1/8"
      - "::1"
access:
  configure:
    admin: allow
access_rules:
  local:
    allow: local
  c2s:
    deny: blocked
    allow: all
  announce:
    allow: admin
  configure:
    allow: admin
  muc_create:
    allow: loopback
  pubsub_createnode:
    allow: local
  trusted_network:
    allow: loopback
api_permissions:
  "console commands":
    from:
      - ejabberd_ctl
    who: all
    what: "*"
  "admin access":
    who:
      - access:
          - allow:
            - acl: admin
      - oauth:
        - scope: "ejabberd:admin"
        - access:
          - allow:
              - acl: admin
    what:
      - "*"
      - "!stop"
      - "!start"
  "public commands":
    who: all
    what:
      - "status"
      - "register"
      - "connected_users_number"
      - "create_room"
      - "set_room_affiliation"
      - "subscribe_room"
      - "unregister"
      - "unsubscribe_room"
      - "create_room_with_opts"
      - "send_stanza"
shaper:
  normal: 1000
  fast: 50000
shaper_rules:
  max_user_sessions: 10
  max_user_offline_messages:
    5000: admin
    100: all
  c2s_shaper:
    none: admin
    normal: all
  s2s_shaper: fast
modules:
  mod_adhoc: {}
  mod_admin_extra: {}
  mod_announce:
    access: announce
  mod_avatar: {}
  mod_blocking: {}
  mod_bosh: {}
  mod_caps: {}
  mod_carboncopy: {}
  mod_client_state: {}
  mod_configure: {}
  mod_disco: {}
  mod_fail2ban: {}
  mod_http_api: {}
  mod_http_upload:
    put_url: https://@HOST@:5443/upload
  mod_last: {}
  mod_mam:
    ## Mnesia is limited to 2GB, better to use an SQL backend
    ## For small servers SQLite is a good fit and is very easy
    ## to configure. Uncomment this when you have SQL configured:
    db_type: sql
    assume_mam_usage: true
    default: always
  mod_mqtt: {}
  mod_muc:
    access:
      - allow
    access_admin:
      - allow: admin
    access_create: muc_create
    access_persistent: muc_create
    access_mam:
      - allow
    default_room_options:
      mam: true
      persistent: true
      allow_subscription: true
    db_type: sql
    preload_rooms: false
  mod_muc_admin: {}
  mod_offline:
    store_empty_body: unless_chat_state
    store_groupchat: true
    access_max_user_messages: max_user_offline_messages
  mod_ping:
    send_pings: true
    ping_interval: 5
    ping_ack_timeout: 5
    timeout_action: kill
  mod_privacy: {}
  mod_private: {}
  mod_proxy65:
    access: local
    max_connections: 5
  mod_pubsub:
    access_createnode: pubsub_createnode
    plugins:
      - flat
      - pep
    force_node_config:
      ## Avoid buggy clients to make their bookmarks public
      storage:bookmarks:
        access_model: whitelist
  mod_push: {}
  mod_push_keepalive: {}
  mod_register:
    ## Only accept registration requests from the "trusted"
    ## network (see access_rules section above).
    ## Think twice before enabling registration from any
    ## address. See the Jabber SPAM Manifesto for details:
    ## https://github.com/ge0rg/jabber-spam-fighting-manifesto
    ip_access: trusted_network
  mod_roster:
    versioning: true
    db_type: sql
  mod_s2s_dialback: {}
  mod_shared_roster: {}
  mod_stream_mgmt:
    resume_timeout: 60 sec
    resend_on_timeout: if_offline
  mod_stun_disco: {}
  mod_vcard: {}
  mod_vcard_xupdate: {}
  mod_version:
    show_os: false
  #mod_cobrowser: {}
  mod_chatevents:
    auth_token: "secret"
    post_url: "http://my-service-development-960667868.ap-southeast-1.elb.amazonaws.com:9050/chat-service/pushChatEvent?client_id=communication-gateway-service"
    confidential: true
  mod_recommendation:
    auth_token: "secret"
    post_url: "http://my-service-development-960667868.ap-southeast-1.elb.amazonaws.com:9050/chat-service/sendRecommendations?client_id=communication-gateway-service"
    confidential: true
  mod_offline_http_post:
    auth_token: "secret"
    post_url: "http://my-service-development-960667868.ap-southeast-1.elb.amazonaws.com:9050/chat-service/handleChatEvent?client_id=communication-gateway-service"
    confidential: true
  mod_read_markers:
    auth_token: "secret"
    post_url: "http://my-service-development-960667868.ap-southeast-1.elb.amazonaws.com:9050/chat-service/handleChatEvent?client_id=communication-gateway-service"
    confidential: true
sql_type: mysql
sql_server: "dev-databases.mycompany.com"
sql_database: "ejabberd"
sql_username: "dev_mysql_rw"
sql_password: "6ZBexBTjdTNCyb"
sql_port: 3306
default_db: sql
auth_method: sql
### Local Variables:
### mode: yaml
### End:

...
```
</details>

## Errors from error.log/crash.log

400 - "The room does not exist"

## Bug description

1. Created muc_room with a participant using API - /api/create_room, /api/subscribe_room, /api/set_room_affiliation
2. Restarted ejabberd server
3. Tried adding another user to the same room using API -  /api/subscribe_room

ERROR get thrown - 400 "The room does not exist"

This is not a problem when preload_room is set to true. However, we cannot be using that in order to minimise infra costs.

**Comment by badlop** [comment]
You didn't mention what ejabberd version you are using. I cannot reproduce that problem using ejabberd from git:

I install ejabberd from git, using the default configuration, then set for mod_muc:
```yaml
  preload_rooms: false
```

Now start ejabberd for the first time, setup the accounts, the room, stop and start ejabberd:
```
$ ejabberdctl start

$ ejabberdctl register admin localhost asd
User admin@localhost successfully registered

$ ejabberdctl register user2 localhost asd
User user2@localhost successfully registered

$ ejabberdctl create_room room1 conference.localhost localhost
$ ejabberdctl change_room_option room1 conference.localhost persistent true
$ ejabberdctl change_room_option room1 conference.localhost allow_subscription true

$ ejabberdctl subscribe_room admin@localhost AdMIN room1@conference.localhost "urn:xmpp:mucsub:nodes:messages"
urn:xmpp:mucsub:nodes:messages

$ ejabberdctl stop
$ ejabberdctl start
```

I check in the erlang process list that no muc room process was created
```
$ ejabberdctl subscribe_room user2@localhost UsER2 room1@conference.localhost "urn:xmpp:mucsub:nodes:messages"
urn:xmpp:mucsub:nodes:messages
```

And I see the muc room process was just created

Login with user2, and open a XML console in the client. Then send a message to the room:
```
ejabberdctl send_message groupchat admin@localhost room1@conference.localhost "" Holaa
```

User2 receives the message correctly:
<details>

```xml
<message to='user2@localhost'
	from='room1@conference.localhost'
	id='1649233969293802'>
  <archived by='user2@localhost'
	id='1649233969294350'
	xmlns='urn:xmpp:mam:tmp'/>
  <stanza-id by='user2@localhost'
	id='1649233969294350'
	xmlns='urn:xmpp:sid:0'/>
  <event xmlns='http://jabber.org/protocol/pubsub#event'>
    <items node='urn:xmpp:mucsub:nodes:messages'>
      <item id='1649233969293802'>
        <message to='user2@localhost'
	from='room1@conference.localhost/AdMIN'
	type='groupchat'
	id='13684065620118385089'
	xmlns='jabber:client'>
          <archived by='room1@conference.localhost'
	id='1649233969293802'
	xmlns='urn:xmpp:mam:tmp'/>
          <stanza-id by='room1@conference.localhost'
	id='1649233969293802'
	xmlns='urn:xmpp:sid:0'/>
          <body>Holaa</body>
        </message>
      </item>
    </items>
  </event>
</message>
```
</details>

**Comment by arjun9** [comment]
[badlop](https://github.com/badlop): ejabberd version: 20.04

**Comment by arjun9** [comment]
Hi @badlop,
Can this diff simply be patched to 20.04 without any issue ?

**Comment by badlop** [comment]
The patch applies cleanly to 20.04, but it includes several changes... The good thing is that it only affects mod_muc_admin commands, does not affect the database. If it doesn't work correctly, probably you can revert it, restart ejabberd and continue with the old code.
